pkgpanda/util.py

```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
     return session
 
 
+def stream_remote_file_with_retries(out_filename, url, retries=4):
+    def cleanup():
+        try:
+            os.remove(out_filename)
+        except Exception:
+            pass
+
+    while True:
+        with open(out_filename, "w+b") as f:
+            r = get_requests_retry_session().get(url, stream=True)
+            if r.status_code == 301:
+                raise Exception("got a 301")
+            r.raise_for_status()
+
+            for chunk in r.iter_content(chunk_size=4096):
+                f.write(chunk)
+            f.flush()
+
+            content_length = int(r.headers['content-length'])
+            final_size = os.fstat(f.fileno()).st_size
+            if final_size == content_length:
+                return r
+            else:
+                f.close()
```

Comment:
Calling `f.close()` is often a sign that you could restructure your code, as the context manager `with open(...) as f:` will do that automatically when you leave its scope.

Reply:
Ah, yes, I wanted to clean this up after confirming general approach. Let's see if we can get rid of a lot of it when using `retrying`...